### PR TITLE
LibX86+disasm: Support x86-64 (attempt #2)

### PR DESCRIPTION
**Pull request by supercomputer7**
This is a rebased branch of #13300.

Still need to look into the comments of what the original author had to do and fix to ensure it's ready for merging.

### CODE REVIEW

**Comment by supercomputer7** [comment]
> With #15467 around the corner we might be able to remove the 16 bit and 32 bit tables in a future PR.

I had a discussion about this quite a long time ago, but it might be valuable to keep 16 bit and 32 bit support for the x86 platform, especially because I don't want the Kernel to have such code to emulate BIOS calls, but to let userspace to deal with this.
This PR also raises the question - since aarch64 port is quite close, what will happen then? Will UserspaceEmulator will need a LibAArch64 now? that seems tedious to me.
Besides finding memory leaks (as far as I remember this is the purpose of this program) what else it does for us currently?

**Comment by Hendiadyoin1** [comment]
It also detects accesses to out of bounds memory regions, as well as some UB checks

**Comment by ADKaster** [comment]
It's the equivalent of the memcheck tool from valgrind. With the option to add other tools in the future. 

I think we should at least have it working on one architecture, I don't think it's a requirement to port it to every arch. But possibly, yes. Technically with this we could maybe run x86_64 serenity userspace apps on aarch64 serenity, like userspace emulation in qemu.

**Comment by supercomputer7** [comment]
> It's the equivalent of the memcheck tool from valgrind. With the option to add other tools in the future.
> 
> I think we should at least have it working on one architecture, I don't think it's a requirement to port it to every arch. But possibly, yes. Technically with this we could maybe run x86_64 serenity userspace apps on aarch64 serenity, like userspace emulation in qemu.

Although the idea of running x86-64 binaries on aarch64 CPU sounds quite nice, I'd advocate against it unless there's a proper way to contain the x86-64 userland somehow so we don't have a mess. Something like simple `chroot` functionality can do that.

Anyway, I need to find time to finish this. There's one comment left to address in this PR, but I didn't look in the old PR this is based upon, so I might need to work on this a bit more.

File: Userland/Libraries/LibX86/Instruction.cpp [hunk 1]
```diff
@@ -1371,79 +1371,35 @@ String MemoryOrRegisterReference::to_string_a16() const
...
-        break;
-    case 0xC0:
-        scale = "*8";
-        break;
-    }
-    switch ((sib >> 3) & 0x07) {
-    case 0:
-        index = "eax";
+    switch (m_sib_scale) {
+    case 0:;
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibX86/Instruction.cpp:1444]: I find that extra `;` weird

File: Userland/Libraries/LibX86/Instruction.h [hunk 2]
```diff
@@ -518,8 +525,12 @@ class MemoryOrRegisterReference {
         u16 m_displacement16;
     };
 
-    u8 m_rm_byte { 0 };
-    u8 m_sib { 0 };
+    u8 m_mod : 2 { 0 };
+    u8 m_reg : 4 { 0 };
+    u8 m_rm : 4 { 0 };
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibX86/Instruction.h:609]: At this point it should technically still be 3 each and 4 only with the next commit,
But you don't need to change it, if you don't want

Comment (Hendiadyoin1) [reply Userland/Libraries/LibX86/Instruction.h:609]: (Likely same for SIB, but haven't read that part of the manual)

**Review by Hendiadyoin1 — COMMENTED** [review]
Other than what I found, which aren't severe, I think it is good to go,
Checking if `Build/lagom/disasm Build/lagom/js` or the like work, would serve as a good verification.

**Review by ADKaster — APPROVED** [review]
Looks good to me mod Hendiadyoin's comments

**Comment by awesomekling** [comment]
Let's just bring this in and continue iterating in-tree.

**Pull request merged**